lib/vm/src/instance.rs

```diff
@@ -688,6 +688,189 @@ impl Instance {
     }
 }
 
+/// This is an intermediate type that manages the raw allocation and
+/// metadata when creating an [`Instance`].
+///
+/// This type will free the allocated memory if it's dropped before
+/// being used.
+///
+/// The [`UnpreparedInstance::instance_layout`] computes the correct
+/// layout to represent the wanted [`Instance`].
+///
+/// Then we use this layout to allocate an empty `Instance` properly.
+pub struct UnpreparedInstance {
+    /// The buffer that will contain the [`Instance`] and dynamic fields.
+    instance_ptr: NonNull<Instance>,
+    /// The layout of the `instance_ptr` buffer.
+    instance_layout: Layout,
+    /// Information about the offsets into the `instance_ptr` buffer for
+    /// the dynamic fields.
+    offsets: VMOffsets,
+    /// Whether or not this type has transferred ownership of the
+    /// `instance_ptr` buffer. If it has not when being dropped,
+    /// the buffer should be freed.
+    consumed: bool,
+}
+
+impl Drop for UnpreparedInstance {
+    fn drop(&mut self) {
+        if !self.consumed {
+            // If `consumed` has not been set, then we still have ownership
+            // over the buffer and must free it.
+            let instance_ptr = self.instance_ptr.as_ptr();
+            unsafe {
+                std::alloc::dealloc(instance_ptr as *mut u8, self.instance_layout);
+            }
+        }
+    }
+}
+
+impl UnpreparedInstance {
+    /// Allocates instance data for use with [`InstanceHandle::new`].
+    ///
+    /// Returns a wrapper type around the allocation and 2 vectors of
+    /// pointers into the allocated buffer. These lists of pointers
+    /// correspond to the location in memory for the local memories and
+    /// tables respectively. These pointers should be written to before
+    /// calling [`InstanceHandle::new`].
+    pub fn new(
+        module: &ModuleInfo,
+    ) -> (
+        UnpreparedInstance,
+        Vec<NonNull<VMMemoryDefinition>>,
+        Vec<NonNull<VMTableDefinition>>,
+    ) {
+        let offsets = VMOffsets::new(mem::size_of::<*const u8>() as u8, module);
```

Comment:
Side note: I've noticed this `mem::size_of::<*const u8>() as u8` and I wanted to simplify it to `mem::size_of::<usize>() as u8`. We compute the size of a pointer here, and `usize` is exactly that.

That's not a mandatory feedback though 😃.